File: Example/PDStrategy/PDStrategy/PDStrategy/Models/PDController.m
//
//  PDController.m
//  PDStrategy
//
//  Created by Pavel Deminov on 19/11/2017.
//  Copyright © 2017 Pavel Deminov. All rights reserved.
//

#import "PDController.h"
#import "PDSection.h"

@implementation PDController

- (instancetype)init {
    self = [super init];
    if (self) {
        [self setup];
    }
    return self;
}

- (void)setup {
    
}

- (void)appendData:(nullable id)data {
    
}

- (void)validate {
    ValidationState state = ValidationStateNone;
    for (PDSection *section in self.sections) {
        [section validate];
        state = section.state;
        if (state == ValidationStateInvalid) {
            self.errorContainer = section;
            break;
        }
    }
    self.state = state;
}

- (void)invalidate {
    self.state = ValidationStateNone;
    self.errorContainer = nil;
}

- (NSDictionary *)dictionary {
    NSMutableDictionary *dict = [NSMutableDictionary new];
    for (PDSection *container in self.sections) {
        NSDictionary *containerDict = [container dictionary];
        [dict addEntriesFromDictionary:containerDict];
    }
    return dict;
}

- (nullable NSIndexPath *)errorIndexPath {
    if (!self.errorContainer) {
        return nil;
    }
    
    NSInteger sectionIndex = [self.sections indexOfObject:self.errorContainer];
    PDSection *section = self.errorContainer;
    NSInteger rowindex = [self.errorContainer.items indexOfObject:section.errorItem];
    
    return [NSIndexPath indexPathForRow:rowindex inSection:sectionIndex];
}

- (nullable NSIndexPath *)indexPathForItemInfo:(id <PDItemInfo>)itemInfo {
    
    for (id <PDSectionInfo> sectionInfo in self.sections) {
        for (id <PDItemInfo> item in sectionInfo.items) {
            if (itemInfo == item) {
                NSInteger section = [self.sections indexOfObject:sectionInfo];
                NSInteger row = [sectionInfo.items indexOfObject:item];
                return [NSIndexPath indexPathForRow:row inSection:section];
            }
        }
    }
    
    return nil;
}

@end

File: Example/PDStrategy/PDStrategy/PDStrategy/Models/PDController.swift
//
//  PDController.swift
//  PDStrategy
//
//  Created by Pavel Deminov on 04/11/2017.
//  Copyright © 2017 Pavel Deminov. All rights reserved.
//

import UIKit

class PDController: NSObject {
    var sections: [PDSection]?
    var errorSection: PDSection?
    var validationState = PDValidationState.none
    
    required override init() {
        super.init()
        self.setup()
    }
    
    func setup() {
        
    }
    
    func appendData(data: Any) {
        
    }
    
    func validate() {
        validationState = .valid
        
        guard let sections = sections else {
            return
        }
        
        for section in sections {
            section.validate()
            validationState = section.validationState
            if validationState == .invalid {
                errorSection = section
                break
            }
        }
    }
    
    func invalidate() {
        validationState = PDValidationState.none
        errorSection = nil
    }
    
    func dictionary() -> [AnyHashable: Any] {
        var dict = [AnyHashable: Any]()
        
        guard let sections = sections else {
            return dict
        }
        
        for section: PDSection in sections {
            let sectionDict = section.dictionary()
            for (k, v) in sectionDict { dict.updateValue(v, forKey: k) }
        }
        return dict
    }
    
    func errorIndexPath() -> IndexPath? {
        
        weak var wSelf = self
        guard let  sectionIndex = sections?.index(where: { (section) -> Bool in
            return section == wSelf?.errorSection
        }) else {
            return nil
        }
        
        guard let  rowIndex = errorSection?.items?.index(where: { (item) -> Bool in
            return item.itemHash == wSelf?.errorSection?.errorItem?.itemHash
        }) else {
            return nil
        }
    
        return IndexPath(row: rowIndex, section: sectionIndex)
    }
    
}
